### PR TITLE
Add verbatim identifiers

### LINKED ISSUE #25: Verbatim identifiers

### LINKED ISSUE #46: syntax highlight is dead after '@' keyword.
_From @sassembla on March 5, 2017 9:23_

## Environment data
`dotnet --info` output:
.NET Command Line Tools (1.0.0-preview2-003148)

Product Information:
 Version:            1.0.0-preview2-003148
 Commit SHA-1 hash:  18ca244551

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.12
 OS Platform: Darwin
 RID:         osx.10.11-x64

VS Code version:version 1.10.1 (1.10.1)
C# Extension version:1.7.0

## Steps to reproduce
use '@' character as field name like below.

```
public class SampleClass {
    public int MyInt;

    public int @class;// syntax highlight is disabled from @class.

    public double MyDouble;

    public void SomeMethod () {        
        var str = "";// syntax highlight is back!
    }
}
```

## Expected  behavior
syntax highlight is alive.

## Actual behavior
syntax highlight is dead after '@' keyword.

_Copied from original issue: OmniSharp/omnisharp-vscode#1288_

### PR DESCRIPTION
**Pull request by Muchiachio**
Sorry for a long diff, but had to add a lot of tests to make sure everything works.
Resolves #25.
Fixes #46.

### CODE REVIEW

File: src/csharp.tmLanguage.yml [hunk 1]
```diff
@@ -410,7 +410,7 @@ repository:
     - match: \b(in|out)\b
       captures:
         '1': { name: storage.modifier.cs }
-    - match: \b([_[:alpha:]][_[:alnum:]]*)\b
+    - match: (@?\b[_[:alpha:]][_[:alnum:]]*)\b
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Shouldn't `@?` and `\b` be swapped?

File: src/csharp.tmLanguage.yml [hunk 2]
```diff
@@ -718,10 +718,10 @@ repository:
     - include: '#block'
 
   constructor-declaration:
-    begin: (?=[_[:alpha:]][_[:alnum:]]*\s*\()
+    begin: (?=@?[_[:alpha:]][_[:alnum:]]*\s*\()
     end: (?<=\})|(?=;)
     patterns:
-    - match: \b([_[:alpha:]][_[:alnum:]]*)\b
+    - match: (@?\b[_[:alpha:]][_[:alnum:]]*)\b
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Shouldn't `@?` and `\b` be swapped here?

File: src/csharp.tmLanguage.yml [hunk 3]
```diff
@@ -1157,7 +1157,7 @@ repository:
               (?:\s*\[(?:\s*,\s*)*\]\s*)* # array suffix?
             )
           )\s*
-          (?:\b(\g<identifier>)\b)?
+          (?:(\g<identifier>)\b)?
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Why did you remove this `\b`?

File: src/csharp.tmLanguage.yml [hunk 4]
```diff
@@ -1360,7 +1360,7 @@ repository:
     - include: '#punctuation-comma'
     - match: |-
         (?x) # e.g. x
-        \b([_[:alpha:]][_[:alnum:]]*)\b\s*
+        (@?\b[_[:alpha:]][_[:alnum:]]*)\b\s*
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Should `@?` and `\b` be swapped?

File: src/csharp.tmLanguage.yml [hunk 5]
```diff
@@ -1379,7 +1379,7 @@ repository:
...
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Should `@?` and `\b` be swapped?

File: src/csharp.tmLanguage.yml [hunk 6]
```diff
@@ -1405,7 +1405,7 @@ repository:
           )
         )
       )\s+
-      \b(\g<identifier>)\b\s*
+      (\g<identifier>)\b\s*
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Why remove `\b` here?

File: src/csharp.tmLanguage.yml [hunk 7]
```diff
@@ -1439,7 +1439,7 @@ repository:
...
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Why remove `\b` here?

File: src/csharp.tmLanguage.yml [hunk 8]
```diff
@@ -2092,7 +2092,7 @@ repository:
           (?:\s*\[(?:\s*,\s*)*\]\s*)* # array suffix?
         )
       )?
-      \b(\g<identifier>)\b\s*
+      \s+(\g<identifier>)\b\s*
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: This doesn't feel quite right. Why was this change necessary?

File: src/csharp.tmLanguage.yml [hunk 9]
```diff
@@ -2123,7 +2123,7 @@ repository:
     begin: |-
       (?x)
       \b(let)\b\s*
-      \b([_[:alpha:]][_[:alnum:]]*)\b\s*
+      (@?\b[_[:alpha:]][_[:alnum:]]*)\b\s*
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Shouldn't `@?` and `\b` be swapped?

File: src/csharp.tmLanguage.yml [hunk 10]
```diff
@@ -2164,7 +2164,7 @@ repository:
           (?:\s*\[(?:\s*,\s*)*\]\s*)* # array suffix?
         )
       )?
-      \b(\g<identifier>)\b\s*
+      \s+(\g<identifier>)\b\s*
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Ditto here. I'm not sure why this is necessary here. I would expect this to be after the type.

File: src/csharp.tmLanguage.yml [hunk 11]
```diff
@@ -2199,7 +2199,7 @@ repository:
     match: |-
       (?x)
       \b(into)\b\s*
-      \b([_[:alpha:]][_[:alnum:]]*)\b\s*
+      (@?\b[_[:alpha:]][_[:alnum:]]*)\b\s*
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Confused by position of `@?` and `\b'. Doesn't that defeat the purpose of `\b'?

File: src/csharp.tmLanguage.yml [hunk 12]
```diff
@@ -2250,7 +2250,7 @@ repository:
...
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Ditto here

File: src/csharp.tmLanguage.yml [hunk 13]
```diff
@@ -2260,7 +2260,7 @@ repository:
     - begin: |-
         (?x)
         (?:\b(async)\b\s*)?
-        \b([_[:alpha:]][_[:alnum:]]*)\b\s*
+        (@?\b[_[:alpha:]][_[:alnum:]]*)\b\s*
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: And here

File: src/csharp.tmLanguage.yml [hunk 14]
```diff
@@ -2332,7 +2332,7 @@ repository:
           (?:\s*\[(?:\s*,\s*)*\]\s*)* # array suffix?
         )
       )?
-      \b(\g<identifier>)\b\s*
+      (\g<identifier>)\b\s*
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Why did the `\b` need to be removed here?

File: src/csharp.tmLanguage.yml [hunk 15]
```diff
@@ -2390,7 +2390,7 @@ repository:
           (?:\s*\[(?:\s*,\s*)*\]\s*)* # array suffix?
         )
       )
-      (?:\b(?<tuple-name>\g<identifier>)\b)?
+      (?:(?<tuple-name>\g<identifier>)\b)?
```
Comment (DustinCampbell) [inline src/csharp.tmLanguage.yml]: Why did the `\b` need to be removed here?

**Comment by Muchiachio** [comment]
`\b'` swaps are needed, because `@` is not considered as a part of a word and thus doesn't match a word boundary, so for example `@my_var` doesn't get matched with `\b@my_var\b`.

As for `\b` removals, it has a similar problem because of `\b(\g<identifier>)`, it gets mapped to
`\b(\g@?[_[:alpha:]][_[:alnum:]]*)`, which again can't be matched because of a leading `@`. Also all the patterns which used the removed `\b` are followed by leading words separated by space so they will have a correct leading boundary anyway, that's why I removed them.

I was also trying to replace those few `<identifier>` patterns with `@?[_[:alpha:]][_[:alnum:]]*`, but it didn't work for some reason.

File: src/csharp.tmLanguage.yml (hunk 10 above)
Comment (Muchiachio) [reply src/csharp.tmLanguage.yml]: After the type spaces get consumed by the type pattern, but the optional pattern (one line above) doesn't ensure a space needed to create a correct word boundary for `@`.

**Comment by DustinCampbell** [comment]
It does make sense that `\b` would get in the way once `@` can appear at the front of an identifier match. But why the swaps? Wouldn't it be better to just remove the `\b` in those cases?

**Comment by Muchiachio** [comment]
Yes, in this case they can be removed, just left them there because they were there.
Should I do that?

**Comment by Muchiachio** [comment]
Removed.
